### PR TITLE
Fix Cubic Nether Generation

### PR DESCRIPTION
**Pull request by CorgiTaco**
-Adds Nether Aquifers
-Fixes all nether features to work in a cubic context

As it stands it works for the *most* part, still working out the bugs.

### CODE REVIEW

**Comment by CorgiTaco** [comment]
Ready for review :D

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/gen/feature/LavaLeakFix.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            for (int localY = 0; localY < IBigCube.DIAMETER_IN_BLOCKS; localY++) {
+                for (int localZ = 0; localZ < IBigCube.DIAMETER_IN_BLOCKS; localZ++) {
+                    mutable.set(localX, localY, localZ);
+                    if (cube.getBlockState(localX, localY, localZ).getBlock() != Blocks.LAVA) {
+                        continue;
+                    }
+                    for (Direction direction : Direction.values()) {
+                        if (direction == Direction.UP) {
+                            continue;
+                        }
```
Comment (alcatrazEscapee) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/gen/feature/LavaLeakFix.java]: `Direction.values()` does a full array allocation every time it's invoked. As a result you're allocating (at default config values) up to 32,768 `Direction[]` arrays.

You could reduce that down to one outside the loop, and even strip off the `UP` direction to reduce the need for the `if (...) continue`

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/gen/feature/LavaLeakFix.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+                        if (level.getBlockState(mutable.set(Coords.localToBlock(cubePos.getX(), localX),
+                            Coords.localToBlock(cubePos.getY(), localY), Coords.localToBlock(cubePos.getZ(), localZ)).move(direction)).isAir()) {
+                            if (direction == Direction.DOWN) {
+                                level.setBlock(mutable, generator.getBaseStoneSource().getBaseBlock(mutable), 2);
+                            } else {
+                                if (random.nextInt(5) == 0) {
+                                    continue;
+                                }
+                                SurfaceBuilderConfiguration surfaceBuilderConfiguration = context.level().getBiome(mutable).getGenerationSettings().getSurfaceBuilder().get().config();
```
Comment (alcatrazEscapee) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/gen/feature/LavaLeakFix.java]: Without profiling it's impossible to tell, but I would *very much guess* that the slowest thing in this method is going to be the `getBiome()` call here. As to my understanding it's at the very least making a biome magnifier / zoomer call, which is not the fastest thing in the world. At worst it's re-generating biomes.

**Review by alcatrazEscapee — COMMENTED** [review]
Corgi asked for performance tips. Obviously, profile it first and foremost but here's some simple first impressions.

**Comment by CorgiTaco** [comment]
Ready for review stinkies

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/MixinPortalForcer.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    @SuppressWarnings("UnresolvedMixinReference")
+    @Redirect(method = "lambda$findPortalAround$5(Lnet/minecraft/world/entity/ai/village/poi/PoiRecord;)Lnet/minecraft/BlockUtil$FoundRectangle;", at = @At(value = "INVOKE",
+        target = "Lnet/minecraft/server/level/ServerChunkCache;addRegionTicket(Lnet/minecraft/server/level/TicketType;Lnet/minecraft/world/level/ChunkPos;ILjava/lang/Object;)V"))
+    private <T> void addCubeRegionTicket(ServerChunkCache serverChunkCache, TicketType<T> ticketType, ChunkPos chunkPos, int radius, T argument, PoiRecord poiRecord) {
+        if (!((CubicLevelHeightAccessor) serverChunkCache.getLevel()).isCubic()) {
+            serverChunkCache.addRegionTicket(ticketType, chunkPos, radius, argument);
+            return;
+        }
+        serverChunkCache.addRegionTicket(ticketType, chunkPos, radius, argument); // TODO: Load order?
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/MixinPortalForcer.java]: Column ticket is no longer needed

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/MixinNaturalSpawner.java [hunk 4]
```diff
@@ -71,18 +76,83 @@ private static void returnOnBrokenPosition(LevelReader reader, EntityType<?> ent
...
+        }
+        return false;
+    }
+
+//    @Redirect(method = "getTopNonCollidingPos", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/LevelReader;getMinBuildHeight()I"))
+//    private static int useCubeMinY(LevelReader levelReader) {
+//        if (!((CubicLevelHeightAccessor) levelReader).isCubic()) {
+//            return levelReader.getMinBuildHeight();
+//        }
+//        return Coords.cubeToMinBlock(((CubeWorldGenRegion) levelReader).getMainCubeY());
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/MixinNaturalSpawner.java]: Why is all this commented out code here?

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]

**Review by Niko-sk2x — APPROVED** [review]

**Pull request merged**